File: ex/writeshmem.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "shared_memory.h"
#include "shared_memory_cnf.h"

#define KEY_ID_ARG_INDEX 2
int main(int argc, char* argv[])
{
    key_t key;
    key = atoi(argv[KEY_ID_ARG_INDEX]);
    printf("Key Id: %d\n",key);

    if (argc != 3)
    {
        printf("usage - %s [stuff to write]",argv[0]);
        return -1;
    }

    // grab the shared memory blick
    char* block = attach_memory_block(key, BLOCK_SIZE);
    if (block == NULL)
    {
        printf("ERROR: couldn't get block\n");
        return -1;
    }

    printf("Writing: \"%s\"\n", argv[1]);
    strncpy(block,argv[1],BLOCK_SIZE);

    detach_memory_block(block);

    return 0;
}

File: ex/writeshmem.py
import sysv_ipc
import shared_memory_cnf as cfg

import argparse
import sys

print ('Number of arguments:', len(sys.argv), 'arguments.')
print ('Argument List:', str(sys.argv))

if (len(sys.argv) != 3):
    print(f"usage - {sys.argv} // no args)")

writeMessage = str(sys.argv[1])
key = int(sys.argv[2])

def main():
    sysv_memory = ''
    print(f"Key Id: {key}\n");
    
    try:
        sysv_memory=sysv_ipc.SharedMemory(key, sysv_ipc.IPC_CREAT, 644, cfg.BLOCK_SIZE)
    except Exception as e:
        print(e)
        print(f"Memmory block with is: {key} cannot be created\n")
    
    try:
        print(f"Writing: \"{writeMessage}\"\n")
        sysv_memory.write(writeMessage + '\0')
    except Exception as e:
        print(e)
        print(f"Write data failed\n")

main()
